src/classes/PS_IntegrationServiceConfig.cls

```diff
@@ -61,17 +85,124 @@ public with sharing class PS_IntegrationServiceConfig {
+    * @return Configuration 
+    */
+    private static Configuration psConfig {
+        get {
+            if (psConfig == null) {
+                psConfig = new Configuration();
+            }
+            return psConfig;
+        } set;
+    }
+
+    /**
+    * @description Inner class responsible for retrieving Payment Services configuration from
+    *              application storage and building and exposing a set of key-value
+    *              configuration pairs.
+    *              We use 'without sharing' to ensure that the query against 
+    *              Payment_Services_Configuration__c always works regardless of what permissions 
+    *              the current user has
+    */
+    @TestVisible
+    public without sharing class Configuration {
+        public Configuration() {}
+
+        // use a class variable to enable caching and 
+        // avoid querying the config object multiple times
+        List<Payment_Services_Configuration__c> lastModifiedRecords;
+
+        /* @description returns a map of key value pairs from the configuration object
+        */
+        public Map<String, String> keyValueMap {
+            get {
+                if (keyValueMap == null) {
+                    keyValueMap = new Map<String, String>();
+                    String makanaKey;
+                    for (Payment_Services_Configuration__c configRecord :
+                            getConfigRecordsByName(
+                                    new List<String> {
+                                            PAYMENTS_SERVICE_NAME, MAKANA_SERVICE_NAME
+                                    }
+                            )) {
+                        if (configRecord.Service__c == PAYMENTS_SERVICE_NAME) {
+                            keyValueMap.put(configRecord.Key__c, configRecord.Value__c);
+                        } else if (configRecord.Key__c == API_KEY &&
+                                configRecord.Service__c == MAKANA_SERVICE_NAME) {
+                            makanaKey = configRecord.Value__c;
+                        }
+                    }
+
+                    // If the payments service config does not have an Api key
+                    // and makana does, use the makana Api key
+                    if (makanaKey != null && keyValueMap.get(API_KEY) == null) {
+                        keyValueMap.put(API_KEY, makanaKey);
+                    }
+                }
+                return keyValueMap;
+            }
+            private set;
+        }
+
+        /* @description Returns the last modified configuration record
+        */
+        public Payment_Services_Configuration__c lastModifiedRecord {
+            get {
+                if (lastModifiedRecords == null) {
+                    lastModifiedRecords = getLastModifiedConfigRecord();
+                    lastModifiedRecord = lastModifiedRecords.isEmpty() ? null : lastModifiedRecords[0];
```

Comment:
@vmartinov-sforg  I think this particular line is the issue from before. If `lastModifiedRecords` is an empty collection, the class var is reset to null which means the getter will requery each time.

Reply:
The issue was that I was checking `lastModifiedRecord ` not `lastModifiedRecords`. `lastModifiedRecords` is a class variable, so we are OK - the query gets executed only once whether we have records or not.